homeassistant/components/mqtt/debug_info.py

```diff
@@ -0,0 +1,166 @@
+"""Helper to handle a set of topics to subscribe to."""
+import asyncio
+from functools import partial, wraps
+import logging
+from typing import Any
+
+from homeassistant.helpers.typing import HomeAssistantType
+
+from .const import ATTR_DISCOVERY_PAYLOAD, ATTR_DISCOVERY_TOPIC
+from .models import MessageCallbackType
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_MQTT_DEBUG_INFO = "mqtt_debug_info"
+STORED_MESSAGES = 10
+
+
+def log_messages(hass: HomeAssistantType, entity_id: str) -> MessageCallbackType:
+    """Wrap an MQTT message callback to support message logging."""
+
+    def _log_message(msg):
+        """Log message."""
+        debug_info = hass.data[DATA_MQTT_DEBUG_INFO]
+        messages = debug_info["entities"][entity_id]["topics"][msg.topic]
+        if len(messages) >= STORED_MESSAGES:
+            messages.pop(0)
```

Comment:
Instead of this, use [a deque with a maxlen](https://docs.python.org/3.8/library/collections.html#collections.deque)

Reply:
I did initally, but a deque can't be serialized as JSON so this was the lazy way out. Will look again.

Reply:
We are more often adding messages than serializing to JSON.